src/storage/caching_file_system.cpp

```diff
@@ -0,0 +1,415 @@
+			cached_file.Ranges(guard).clear(); // Invalidate entire cache
+		}
+		cached_file.FileSize(guard) = file_handle->GetFileSize();
+		cached_file.LastModified(guard) = last_modified;
+		cached_file.VersionTag(guard) = version_tag;
+		cached_file.CanSeek(guard) = file_handle->CanSeek();
+		cached_file.OnDiskFile(guard) = file_handle->OnDiskFile();
+	}
+	return *file_handle;
+}
+
+BufferHandle CachingFileHandle::Read(data_ptr_t &buffer, const idx_t nr_bytes, const idx_t location) {
+	BufferHandle result;
+	if (!caching_file_system.enable) {
+		result = caching_file_system.buffer_manager.Allocate(MemoryTag::EXTERNAL_FILE_CACHE, nr_bytes);
+		buffer = result.Ptr();
+		GetFileHandle().Read(buffer, nr_bytes, location);
+		return result;
+	}
+
+	// Get read lock for cached ranges
+	auto guard = cached_file.lock.GetSharedLock();
+	auto &ranges = cached_file.Ranges(guard);
+
+	// First, try to see if we've read from the exact same location before
+	auto it = ranges.find(location);
+	if (it != ranges.end()) {
+		// We have read from the exact same location before
+		if (it->second->GetOverlap(nr_bytes, location) == CachedFileRangeOverlap::FULL) {
+			// The file range contains the requested file range
+			result = TryReadFromFileRange(guard, *it->second, buffer, nr_bytes, location);
+			if (result.IsValid()) {
+				return result;
+			}
+		}
+	}
+
+	// Second, loop through file ranges (ordered by location) to see if any contain the requested file range
+	const auto this_end = location + nr_bytes;
+	vector<shared_ptr<CachedFileRange>> overlapping_ranges;
+	for (it = ranges.begin(); it != ranges.end();) {
+		if (it->second->location >= this_end) {
+			// We're past the requested location
+			break;
+		}
+		// Check if the cached range overlaps the requested one
+		switch (it->second->GetOverlap(nr_bytes, location)) {
+		case CachedFileRangeOverlap::NONE:
+			// No overlap at all
+			break;
+		case CachedFileRangeOverlap::PARTIAL:
+			// Partial overlap, store for potential use later
+			overlapping_ranges.push_back(it->second);
+			break;
+		case CachedFileRangeOverlap::FULL:
+			// The file range fully contains the requested file range, if the buffer is still valid we're done
+			result = TryReadFromFileRange(guard, *it->second, buffer, nr_bytes, location);
+			if (result.IsValid()) {
+				return result;
+			}
+			continue;
+		}
+		it++;
+	}
+
+	// We can unlock now because we copied over the ranges to local
+	guard.reset();
+
+	// Finally, if we weren't able to find the file range in the cache, we have to create a new file range
+	result = caching_file_system.buffer_manager.Allocate(MemoryTag::EXTERNAL_FILE_CACHE, nr_bytes);
+	auto new_file_range = make_shared_ptr<CachedFileRange>(result.GetBlockHandle(), nr_bytes, location, version_tag);
+	buffer = result.Ptr();
+
+	// Interleave reading and copying from cached buffers
+	idx_t current_location = location;
+	idx_t remaining_bytes = nr_bytes;
+	for (auto &overlapping_range : overlapping_ranges) {
```

Comment:
This might cause us to split a single read into multiple read calls - which could potentially degrade performance for remote files especially. Maybe for remote files we only want to do partial range re-use if we are not increasing the number of requests in order to do so (but e.g. only reducing the size of the request).